File: aplicacao-padrao/notificacao.py
class InterfaceNotificacao:
    def __init__(self, tipo, mensagem) -> None:
        pass

    def set_tipo(self, tipo: str) -> None:
        pass

    def set_mensagem(self, mensagem: str) -> None:
        pass

    def get_tipo(self) -> str:
        pass

    def get_mensagem(self) -> str:
        pass


class Notificacao(InterfaceNotificacao):
    def __init__(self, tipo, mensagem) -> None:
        self.tipo = tipo
        self.mensagem = mensagem

    def set_tipo(self, tipo: str) -> None:
        self.set_tipo(tipo)

    def set_mensagem(self, mensagem: str) -> None:
        self.set_mensagem(mensagem)

    def get_tipo(self) -> str:
        return self.tipo

    def get_mensagem(self) -> str:
        return self.mensagem


class Decorador(InterfaceNotificacao):
    def __init__(self, componenteNotificacao: InterfaceNotificacao) -> None:
        self.__componente = componenteNotificacao

    def set_tipo(self, tipo: str) -> None:
        self.__componente.set_tipo(tipo)

    def set_mensagem(self, mensagem: str) -> None:
        self.__componente.set_mensagem(mensagem)

    def get_tipo(self) -> str:
        return self.__componente.get_tipo()

    def get_mensagem(self) -> str:
        return self.__componente.get_mensagem()


class DecoradorNotificacaoFacebook(Decorador):
    def enviar_notificacao(self) -> None:
        print(f"[+] Enviado notificação via Facebook: [{super().get_tipo()}] {super().get_mensagem()}")


class DecoradorNotificacaoSMS(Decorador):
    def enviar_notificacao(self) -> None:
        print(f"[+] Enviado notificação via SMS: [{super().get_tipo()}] {super().get_mensagem()}")


class DecoradorNotificacaoSlack(Decorador):
    def enviar_notificacao(self) -> None:
        print(f"[+] Enviado notificação via Slack: [{super().get_tipo()}] {super().get_mensagem()}")

File: exemplo-notificacao-nodejs/src/notificacao.ts
interface InterfaceNotificacao {
	// Métodos
	setTipo(tipo: string): void;
	setMensagem(mensagem: string): void;
	getTipo(): string;
	getMensagem(): string;
}

export class Notificacao implements InterfaceNotificacao {
	tipo: string;
	mensagem: string;

	constructor(tipo: string, mensagem: string) {
		this.tipo = tipo;
		this.mensagem = mensagem;
	}

	setTipo(tipo: string): void {
		this.tipo = tipo;
	}

	setMensagem(mensagem: string): void {
		this.mensagem = mensagem;
	}

	getTipo(): string {
		return this.tipo;
	}

	getMensagem(): string {
		return this.mensagem;
	}
}

export class Decorador implements InterfaceNotificacao {
	mensagem: string;
	tipo: string;
	
	protected _wrappe: Notificacao;

	constructor(notificacao: Notificacao) {
		this._wrappe = notificacao;
	}

	setTipo(tipo: string): void {
		this._wrappe.setTipo(tipo)
	}

	setMensagem(mensagem: string): void {
		this._wrappe.setMensagem(mensagem)
	}

	getTipo(): string {
		return this._wrappe.getTipo()
	}

	getMensagem(): string {
		return this._wrappe.getMensagem()
	}
}

export class DecoradorNotificacaoFacebook extends Decorador {
	enviarNotificacao(): void {
		console.log(`[+] Enviado notificação via Facebook: [${this.getTipo()}] ${this.getMensagem()}`);
	}
}

export class DecoradorNotificacaoSlack extends Decorador {
	enviarNotificacao(): void {
		console.log(`[+] Enviado notificação via Slack: [${this.getTipo()}] ${this.getMensagem()}`);
	}
}

export class DecoradorNotificacaoSMS extends Decorador {
	enviarNotificacao(): void {
		console.log(`[+] Enviado notificação via SMS: [${this.getTipo()}] ${this.getMensagem()}`);
	}
}